File: client/src/permissions.ts
// Permission flag definitions for the role editor UI. These mirror the
// flags in server/internal/permissions/permissions.go. The Administrator
// flag is a wildcard that grants every other flag.
export interface PermissionDef {
    bit: number
    name: string
    label: string
    group: string
}

export const PERMISSIONS: PermissionDef[] = [
    // Content: Moments
    { bit: 0, name: 'VIEW_MOMENTS', label: 'View Moments', group: 'Moments' },
    { bit: 1, name: 'CREATE_MOMENT', label: 'Create Moment', group: 'Moments' },
    { bit: 2, name: 'EDIT_OWN_MOMENT', label: 'Edit Own Moment', group: 'Moments' },
    { bit: 3, name: 'EDIT_ANY_MOMENT', label: 'Edit Any Moment', group: 'Moments' },
    { bit: 4, name: 'DELETE_OWN_MOMENT', label: 'Delete Own Moment', group: 'Moments' },
    { bit: 5, name: 'DELETE_ANY_MOMENT', label: 'Delete Any Moment', group: 'Moments' },

    // Content: Archives & Tags
    { bit: 6, name: 'MANAGE_ARCHIVES', label: 'Manage Archives', group: 'Archives & Tags' },
    { bit: 7, name: 'MANAGE_TAGS', label: 'Manage Tags', group: 'Archives & Tags' },

    // Chat
    { bit: 8, name: 'VIEW_CHAT', label: 'View Chat', group: 'Chat' },
    { bit: 9, name: 'SEND_CHAT_MESSAGE', label: 'Send Chat Message', group: 'Chat' },
    { bit: 10, name: 'EDIT_OWN_CHAT_MESSAGE', label: 'Edit Own Chat Message', group: 'Chat' },
    { bit: 11, name: 'DELETE_OWN_CHAT_MESSAGE', label: 'Delete Own Chat Message', group: 'Chat' },
    { bit: 12, name: 'DELETE_ANY_CHAT_MESSAGE', label: 'Delete Any Chat Message', group: 'Chat' },

    // Assets
    { bit: 13, name: 'UPLOAD_ASSET', label: 'Upload Asset', group: 'Assets' },
    { bit: 14, name: 'DELETE_ASSET', label: 'Delete Asset', group: 'Assets' },

    // Administration
    { bit: 15, name: 'MANAGE_USERS', label: 'Manage Users', group: 'Administration' },
    { bit: 16, name: 'MANAGE_ROLES', label: 'Manage Roles', group: 'Administration' },
    { bit: 17, name: 'MANAGE_SERVER', label: 'Manage Server', group: 'Administration' },
    { bit: 18, name: 'VIEW_AUDIT_LOG', label: 'View Audit Log', group: 'Administration' },
    { bit: 19, name: 'ADMINISTRATOR', label: 'Administrator (wildcard)', group: 'Administration' },

    // v2.1 additions (ADR-0013)
    { bit: 20, name: 'PIN_MOMENT', label: 'Pin Moments', group: 'Moments' },
    { bit: 21, name: 'MANAGE_TODOS', label: 'Manage Todos', group: 'Modules' },
    { bit: 22, name: 'MANAGE_CANVAS', label: 'Manage Canvases', group: 'Modules' },
    { bit: 23, name: 'MANAGE_BACKUPS', label: 'Manage Backups', group: 'Administration' },
]

// Named bit constants for gating UI (mirror the server flags).
export const PERM = {
    PIN_MOMENT: 20,
    MANAGE_TODOS: 21,
    MANAGE_CANVAS: 22,
    MANAGE_BACKUPS: 23,
    MANAGE_SERVER: 17,
    MANAGE_TAGS: 7,
} as const

export const PERMISSION_GROUPS = [...new Set(PERMISSIONS.map((p) => p.group))]

export function hasPermission(perms: number, bit: number): boolean {
    // Administrator wildcard (bit 19) grants everything.
    if (perms & (1 << 19)) return true
    return (perms & (1 << bit)) !== 0
}

export function togglePermission(perms: number, bit: number): number {
    return perms ^ (1 << bit)
}

File: server/internal/permissions/permissions.go
// Package permissions defines the 19 fine-grained permission flags used by
// the role system. Flags are stored as a bitmask in a uint32.
//
// A user's effective permissions are the bitwise OR of all their roles'
// permission bitmasks. The ADMINISTRATOR flag is a wildcard that grants
// every other flag: check it first via HasAdministrator().
package permissions

// Flag is a single permission bit.
type Flag uint32

// Permission flags. The order doesn't matter; the values must be unique
// powers of two.
const (
	// Content: Moments
	ViewMoments     Flag = 1 << 0
	CreateMoment    Flag = 1 << 1
	EditOwnMoment   Flag = 1 << 2
	EditAnyMoment   Flag = 1 << 3
	DeleteOwnMoment Flag = 1 << 4
	DeleteAnyMoment Flag = 1 << 5

	// Content: Archives & Tags
	ManageArchives Flag = 1 << 6
	ManageTags     Flag = 1 << 7

	// Chat
	ViewChat             Flag = 1 << 8
	SendChatMessage      Flag = 1 << 9
	EditOwnChatMessage   Flag = 1 << 10
	DeleteOwnChatMessage Flag = 1 << 11
	DeleteAnyChatMessage Flag = 1 << 12

	// Assets
	UploadAsset Flag = 1 << 13
	DeleteAsset Flag = 1 << 14

	// Administration
	ManageUsers   Flag = 1 << 15
	ManageRoles   Flag = 1 << 16
	ManageServer  Flag = 1 << 17
	ViewAuditLog  Flag = 1 << 18
	Administrator Flag = 1 << 19

	// v2.1 additions (ADR-0013). New bits added additively, defaulting to 0
	// on existing roles. Not a width change.
	PinMoment     Flag = 1 << 20 // pin/unpin moments (library-shared)
	ManageTodos   Flag = 1 << 21 // create/edit/delete todo lists & items
	ManageCanvas  Flag = 1 << 22 // create/edit/delete canvases & nodes
	ManageBackups Flag = 1 << 23 // list/create/download/restore backups

	// Projects module. Additive like the v2.1 bits: defaults to 0 on
	// existing roles, so only Admin/Owner (wildcard or AllFlags) have it
	// until a role is edited.
	ManageProjects Flag = 1 << 24 // create/edit/delete projects, stages, cards

	// AllFlags is the union of every flag except Administrator.
	// Administrator is a wildcard, see Has().
	AllFlags Flag = ViewMoments | CreateMoment | EditOwnMoment | EditAnyMoment |
		DeleteOwnMoment | DeleteAnyMoment | ManageArchives | ManageTags |
		ViewChat | SendChatMessage | EditOwnChatMessage | DeleteOwnChatMessage |
		DeleteAnyChatMessage | UploadAsset | DeleteAsset | ManageUsers |
		ManageRoles | ManageServer | ViewAuditLog | PinMoment | ManageTodos |
		ManageCanvas | ManageBackups | ManageProjects
)

// Has reports whether the given permission set grants the requested flag.
// Administrator is a wildcard: if set, every flag is granted.
func Has(perms Flag, flag Flag) bool {
	if perms&Administrator != 0 {
		return true
	}
	return perms&flag != 0
}

// HasAdministrator reports whether the permission set includes the
// Administrator wildcard flag.
func HasAdministrator(perms Flag) bool {
	return perms&Administrator != 0
}

// Union returns the bitwise OR of all given permission sets. Used to
// compute a user's effective permissions from their roles.
func Union(flags ...Flag) Flag {
	var result Flag
	for _, f := range flags {
		result |= f
	}
	return result
}

// Preset role permission bundles. These define what each built-in role
// grants. The Owner role always includes Administrator and is not editable.

// OwnerPerms grants everything.
const OwnerPerms Flag = AllFlags | Administrator

// AdminPerms grants everything except ManageRoles and Administrator.
const AdminPerms Flag = AllFlags &^ ManageRoles

// EditorPerms is the trusted collaborator role.
const EditorPerms Flag = ViewMoments | CreateMoment | EditOwnMoment |
	DeleteOwnMoment | ManageArchives | ManageTags | ViewChat |
	SendChatMessage | EditOwnChatMessage | DeleteOwnChatMessage |
	UploadAsset

// ViewerPerms is the default role every user holds (migration 0008), and so is
// the floor for the whole library: it can see everything and add nothing to it:
// no moments, no archives, no tags, no uploads. Contributing is granted
// deliberately, by assigning Editor or a custom role, rather than arriving with
// the invite.
//
// Chat is the exception, and deliberately so: talking is not the same as
// writing to the library. A member who cannot post a moment can still ask
// about one, which is most of the point of having them in the library at all.
// That includes fixing a typo or retracting what they just said: being able to
// post but not unpost is not a coherent floor, and it is their own message
// either way. Editing or deleting anyone else's still takes a wider role.
const ViewerPerms Flag = ViewMoments | ViewChat | SendChatMessage | EditOwnChatMessage | DeleteOwnChatMessage
